### ISSUE TITLE
RFC: Electron backend design

### ISSUE THREAD

**Issue by calebho** [SwatPhonLab/UltraTrace#126]
cc: @keggsmurph21 

Since the UI is written in React and our backend is in Python, I think it makes the most sense to have the main Electron process use some Node API to spawn a simple Python web server which serves the stuff we have in Python already. I think Flask will be sufficient for this use case.

To handle the client-server communication, the two ways I'm most familiar with doing this in web apps are REST(-ish) and GraphQL. REST is nice because its very simple to stand up without the need to use heavyweight frameworks. I think this would basically involve writing new endpoint handlers (which is very easy in Flask) and we'd be good to go. On the client side, I think the browser Fetch API would suffice. This approach probably won't scale very well though. 

GraphQL would require a bit more setup as we'd need a Python framework to get the server side working. Fortunately, I think [Graphene](https://graphene-python.org/) is quite mature. On the client side, we can use Relay or Apollo to create components which nicely encapsulate what data they need to be rendered. This approach will scale very nicely as the complexity of the application grows.

I'm slightly biased towards GraphQL because I don't think the boilerplate is *that* bad, but if we really need to hack something together quickly, then a quick and dirty REST is reasonable to me. When that becomes unmanageable we can migrate as needed.

Open to other approaches as well if you have any.

**Comment by kmurphy4** [comment]
I agree that GraphQL would ultimately be a better choice, but my main concern is that I've never used it before. On the other hand, I'm very familiar with RESTful APIs, and I could get that up and running quickly.

I can try to hack something together in Flask in the next few days.  Do we want to spec out the API here?

**Comment by calebho** [comment]
IMO, it's not super difficult to grasp the server-side for GraphQL. To make an analogy to REST:
* REST read endpoints correspond to GraphQL queries
* REST write endpoints correspond to GraphQL mutations

see https://graphql.org/learn/schema/#the-query-and-mutation-types as an example. Consider this snippet from the example:
```graphql
type Query {
  hero(episode: Episode): Character
  droid(id: ID!): Droid
}
```
In GraphQL-speak, `hero` and `droid` are root queries. In REST-speak, this is roughly saying that there are two `GET` endpoints:
1. `/hero` which has an optional request parameter `episode` of type `Episode`. The response is a `Character`.
2. `/droid` which has a required request parameter `id` of type `ID!` (`!` = required). The response is a `Droid`.

Implementing root queries and mutations are a language + framework dependent ordeal. However, in general each GraphQL object type (e.g. `Query`, `Character`, `Droid`) corresponds 1-1 to a class. This is true in Graphene. At a quick glance, the docs for Graphene are quite good. LMK if you're willing to experiment with it.

Of course, if you really prefer REST, I don't have any objections. In that case let's talk about the API here.

**Comment by keggsmurph21** [comment]
See 8ce0ac3

**Comment by mr-martian** [comment]
I left some comments on that commit.
